src/components/functional_module/src/plugin_manager.cc

```diff
@@ -0,0 +1,410 @@
+      msg_method = value["error"]["data"]["method"].asCString();
+    } else {
+      LOG4CXX_WARN(logger_,
+                   "Message with HMI protocol version can not be handled by "
+                   "plugin manager, because required 'method' field was not "
+                   "found, or was containing an invalid string.");
+      return false;
+    }
+
+    return (hmi_subscribers_.find(msg_method) != hmi_subscribers_.end());
+  }
+
+  return false;
+}
+
+void PluginManager::OnServiceStateChanged(ServiceState state) {
+  for (PluginsIterator it = plugins_.begin(); plugins_.end() != it; ++it) {
+    it->second->OnServiceStateChanged(state);
+  }
+}
+
+void PluginManager::OnHMIResponse(application_manager::MessagePtr msg) {
+  // TODO(PV)
+}
+
+void PluginManager::OnError(ModuleObserver::Errors error, ModuleID module_id) {
+  std::string error_string;
+  switch (error) {
+    case ModuleObserver::Errors::OUT_OF_MEMORY:
+      error_string = "Module run out of memory.";
+      break;
+    case ModuleObserver::Errors::FS_FAILURE:
+      error_string = "Plugin failed to run file system operation.";
+      break;
+    default:
+      break;
+  }
+  LOG4CXX_ERROR(logger_,
+                "Error " << error_string << " was received from module "
+                         << module_id);
+  // TODO(PV)
+}
+
+void PluginManager::RemoveAppExtension(uint32_t app_id) {
+  for (PluginsIterator it = plugins_.begin(); plugins_.end() != it; ++it) {
+    it->second->RemoveAppExtension(app_id);
+  }
+}
+
+bool PluginManager::IsAppForPlugins(
+    application_manager::ApplicationSharedPtr app) {
+  DCHECK(app);
+  if (!app) {
+    return false;
+  }
+
+  bool res = false;
+  for (PluginsIterator it = plugins_.begin(); plugins_.end() != it; ++it) {
+    res = res || it->second->IsAppForPlugin(app);
+  }
+  return res;
+}
+
+bool PluginManager::IsAppForPlugin(
+    application_manager::ApplicationSharedPtr app, ModuleID module_id) const {
+  Modules::const_iterator i = plugins_.find(module_id);
+  return i != plugins_.end() ? i->second->IsAppForPlugin(app) : false;
+}
+
+void PluginManager::OnAppHMILevelChanged(
+    application_manager::ApplicationSharedPtr app,
+    mobile_apis::HMILevel::eType old_level) {
+  DCHECK(app);
+  if (!app) {
+    return;
+  }
+  for (PluginsIterator it = plugins_.begin(); plugins_.end() != it; ++it) {
+    if (it->second->IsAppForPlugin(app)) {
+      LOG4CXX_DEBUG(logger_,
+                    "Application " << app->name().AsMBString() << " of plugin "
+                                   << it->second->GetModuleID()
+                                   << " has changed level from " << old_level
+                                   << " to " << app->hmi_level());
+      it->second->OnAppHMILevelChanged(app, old_level);
+    }
+  }
+}
+
+bool PluginManager::CanAppChangeHMILevel(
+    application_manager::ApplicationSharedPtr app,
+    mobile_apis::HMILevel::eType new_level) {
+  DCHECK(app);
```

Comment:
@yang1070 DCHECK macro removed in release build. WE can use there DCHECK_OR_RETURN_VOID